src/main/software/amazon/event/ruler/ComparableNumber.java

```diff
@@ -1,55 +1,97 @@
-    public static final int NIBBLE_SIZE = 4;
+    static final int MAX_DECIMAL_PRECISON = 6;
 
-    // 1111 0000
-    private static final int UPPER_NIBBLE_MASK = 0xF0;
+    public static final BigDecimal TEN_E_SIX = new BigDecimal("1E6");
+    public static final long FIV_BILL_TEN_E_SIX = new BigDecimal(Constants.FIVE_BILLION).multiply(TEN_E_SIX).longValueExact();
 
-    // 0000 1111
-    private static final int LOWER_NIBBLE_MASK = 0x0F;
+    private static final String HEXES = new String(Constants.HEX_DIGITS, StandardCharsets.US_ASCII);
+    public static final int NIBBLE_SIZE = 4;
+    private static final int UPPER_NIBBLE_MASK = 0xF0; // 1111 0000
+    private static final int LOWER_NIBBLE_MASK = 0x0F; // 0000 1111
 
     private ComparableNumber() {
     }
 
-    static String generate(final double f) {
-        if (f < -Constants.FIVE_BILLION || f > Constants.FIVE_BILLION) {
+    /**
+     * Generates a hexadecimal string representation of a given decimal string value,
+     * with a maximum precision of 6 decimal places and a range between -5,000,000,000
+     * and 5,000,000,000 (inclusive).
+     *
+     * @param str the decimal string value to be converted
+     * @return the hexadecimal string representation of the input value
+     * @throws IllegalArgumentException if the input value has more than 6 decimal places
+     *                                  or is outside the allowed range
+     */
+    static String generate(final String str) {
+        final BigDecimal number = getNumber(str).stripTrailingZeros();
+        if (number.scale() > MAX_DECIMAL_PRECISON) {
+            throw new IllegalArgumentException("Only values upto 6 decimals are supported");
+        }
+
+        final long shiftedBySixDecimals = number.multiply(TEN_E_SIX).longValueExact();
+
+        // faster than doing bigDecimal comparisons
+        if (shiftedBySixDecimals < -FIV_BILL_TEN_E_SIX || shiftedBySixDecimals > FIV_BILL_TEN_E_SIX) {
             throw new IllegalArgumentException("Value must be between " + -Constants.FIVE_BILLION +
                     " and " + Constants.FIVE_BILLION + ", inclusive");
         }
-        return toHexStringSkippingFirstByte((long) (TEN_E_SIX * (Constants.FIVE_BILLION + f)));
+
+        return toHexStringSkippingFirstByte(shiftedBySixDecimals + FIV_BILL_TEN_E_SIX);
+    }
+
+    private static BigDecimal getNumber(String str) {
+        try {
+            return JavaBigDecimalParser.parseBigDecimal(str);
+        } catch (NumberFormatException e) {
+            // maybe it is a hex, fall back to using double where precision isn't guaranteed
+            // we keep existing behaviour of ignore after 6 decimal to avoid breaking backward compatibility
+            // as an acceptable trade-off https://github.com/aws/event-ruler/issues/163
+            return new BigDecimal(Double.parseDouble(str)).setScale(MAX_DECIMAL_PRECISON, RoundingMode.DOWN);
+        }
```

Comment:
In this bit, for Hexadecimal floating point literals, we fallback to the older way method of ignoring beyond 6 decimal places because these are extremely rare and there's a decent chance that decimal errors are coming from Java's `Double.parseDouble()`

Reply:
I don't understand "maybe it is a hex". Hex numbers are not legal JSON.  So that comment should probably change.

Quamina's approach is different. In the case where it can't be turned into a comparable number, we leave it exactly as is in the Event. An example are those numbers in CityLots; you only get a match when the rule contains the exact same representation of the number,  byte for byte.

I'm not sure which approach I prefer because I'm not familiar with the kinds of applications where these kind of weird numbers come up. E.g. I have no idea what kind of logic I'd want if I were processing CityLots records.  I can imagine a future feature where you say something like 

```
{"numeric": [ "=", 5.0123456], "precision": 11}
```

But anyhow, LGTM.